Data/ByteString/Base64/Internal.hs

```diff
@@ -172,16 +172,41 @@ joinWith brk@(PS bfp boff blen) every' bs@(PS sfp soff slen)
 -- This function takes the decoding table (for @base64@ or @base64url@) as
 -- the first paramert.
 decodeWithTable :: Padding -> ForeignPtr Word8 -> ByteString -> Either String ByteString
-decodeWithTable padding decodeFP bs
-    | doPad = go (B.append bs (B.replicate drem 0x3d))
-    | drem /= 0 && (not doPad) = Left "invalid padding"
-    | dlen <= 0 = Right B.empty
-    | otherwise = go bs
+decodeWithTable _ _ (PS _ _ 0) = Right B.empty
+decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
+   case padding of
+     Padded
+       | r /= 0 -> err "Base64-encoded bytestring required to be padded"
+       | otherwise -> go bs
+     Don'tCare
+       | r == 0 -> go bs
+       | r == 2 -> go (B.append bs (B.replicate 2 0x3d))
+       | r == 3 -> go (B.append bs (B.replicate 1 0x3d))
+       | otherwise -> err "Base64-encoded bytestring has invalid size"
+     Unpadded
+       | r == 0 -> validateUnpadded (go bs)
+       | r == 2 -> validateUnpadded (go (B.append bs (B.replicate 2 0x3d)))
+       | r == 3 -> validateUnpadded (go (B.append bs (B.replicate 1 0x3d)))
+       | otherwise -> err "Base64-encoded bytestring has invalid size"
```

Comment:
Like wise here, same deal. We do some additionally low-cost validation to make sure the input does not have padding chars at the end.